operator/cmd/mesh/uninstall.go

```diff
@@ -0,0 +1,200 @@
+	rootArgs := &rootArgs{}
+	uiArgs := &uninstallArgs{}
+	uicmd := &cobra.Command{
+		Use:   "uninstall --revision foo or uninstall --filename iop.yaml",
+		Short: "uninstall the control plane by revision",
+		Long:  "The uninstall command uninstall the control plane by revision or IstioOperator CR",
+		Example: `  # uninstall by revision label foo
+  istioctl uninstall --revision foo
+
+  # uninstall by iop file
+  istioctl uninstall -f iop.yaml
+  
+  # uninstall all istio resources
+  istioctl uninstall --purge
+`,
+		Args: func(cmd *cobra.Command, args []string) error {
+			if uiArgs.revision == "" && uiArgs.filename == "" && !uiArgs.purge {
+				return fmt.Errorf("at least one of the --revision, --filename or --purge flags must be set")
+			}
+			if len(args) > 0 {
+				return fmt.Errorf("istioctl uninstall does not take arguments")
+			}
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return uninstall(cmd, rootArgs, uiArgs, logOpts)
+		}}
+	addUninstallFlags(uicmd, uiArgs)
+	return uicmd
+}
+
+// uninstall uninstalls control plane by either pruning by target revision or deleting specified manifests.
+func uninstall(cmd *cobra.Command, rootArgs *rootArgs, uiArgs *uninstallArgs, logOpts *log.Options) error {
+	l := clog.NewConsoleLogger(cmd.OutOrStdout(), cmd.ErrOrStderr(), installerScope)
+	if err := configLogs(logOpts); err != nil {
+		return fmt.Errorf("could not configure logs: %s", err)
+	}
+	client, err := k8sClient(uiArgs.kubeConfigPath, uiArgs.context)
+	if err != nil {
+		return err
+	}
+
+	cache.FlushObjectCaches()
+	opts := &helmreconciler.Options{DryRun: rootArgs.dryRun, Log: l, ProgressLog: progress.NewLog()}
+	h, err := helmreconciler.NewHelmReconciler(client.Controller(), client.RESTConfig(), nil, opts)
+	if err != nil {
+		return fmt.Errorf("failed to create reconciler: %v", err)
+	}
+
+	// If only revision flag is set, we would prune resources by the revision label.
+	// Otherwise we would merge the revision flag and the filename flag and delete resources by generated manifests.
+	if uiArgs.filename == "" {
+		emptyiops := &v1alpha1.IstioOperatorSpec{Profile: "empty", Revision: uiArgs.revision}
+		iop, err := translate.IOPStoIOP(emptyiops, "empty", iopv1alpha1.Namespace(emptyiops))
+		if err != nil {
+			return err
+		}
+		h, err := helmreconciler.NewHelmReconciler(client.Controller(), client.RESTConfig(), iop, opts)
+		if err != nil {
+			return fmt.Errorf("failed to create reconciler: %v", err)
+		}
+		objectsList, tp, err := h.GetPrunedResources(uiArgs.revision, uiArgs.purge)
+		if err != nil {
+			return err
+		}
+		if err := preCheckWarnings(cmd, rootArgs, uiArgs, uiArgs.revision, tp, l); err != nil {
+			return fmt.Errorf("failed to do preuninstall check: %v", err)
+		}
+
+		if err := h.DeleteObjectsList(objectsList); err != nil {
+			return fmt.Errorf("failed to delete control plane resources by revision: %v", err)
+		}
+		opts.ProgressLog.SetState(progress.StateComplete)
```

Comment:
After I ran this, I saw 17 "Removed <thing>" messages then

```
✔ Installation complete                                                                                   
```

It should say "Uninstall complete".  Make the thing being completed (Installation/Uninstall) a parameter to _/Users/snible/go/src/istio.io/istio/operator/pkg/util/progress/progress.go_.

Reply:
ok done